modules/apps/transfer/types/packet.go

```diff
@@ -62,3 +66,95 @@ func (ftpd FungibleTokenPacketData) ValidateBasic() error {
 func (ftpd FungibleTokenPacketData) GetBytes() []byte {
 	return sdk.MustSortJSON(mustProtoMarshalJSON(&ftpd))
 }
+
+/**
+
+ADR-8 CallbackPacketData implementation
+
+FungibleTokenPacketData implements CallbackPacketDataI interface. This will allow middlewares targetting specific VMs
+to retrieve the desired callback addresses for the ICS20 packet on the source and destination chains.
+
+The Memo is used to ensure that the callback is desired by the user. This allows a user to send an ICS20 packet
+to a contract with ADR-8 enabled without automatically triggering the callback logic which may lead to unexpected
+behaviour.
+
+The Memo format is defined like so:
+
+```json
+{
+	// ... other memo fields we don't care about
+	"callbacks": {
+		"src_callback_address": {contractAddrOnSrcChain},
+		"dest_callback_address": {contractAddrOnDestChain},
+		"src_callback_msg": {jsonObjectForSrcChainCallback},
+		"dest_callback_msg": {jsonObjectForDestChainCallback},
+	}
```

Comment:
Is it intentional to omit methods for retrieving the callback msgs from the interface?

Reply:
As I understand it these are the raw json msgs the contract would accept. Furthermore, if we add additional methods for retrieving them then we're parsing JSON twice to extract values. In this case I'd imagine having a single method `GetPacketCallbacks` which returned an object/strongly typed value rather than a string, because a middleware implementing `OnRecvPacket` for example, needs not only the contract address, but also the json msg to perform some action e.g. swap and stake.